opendevin/condenser/condenser.py

```diff
@@ -0,0 +1,147 @@
+from opendevin.core.exceptions import (
+    SummarizeError,
+)
+from opendevin.core.logger import opendevin_logger as logger
+from opendevin.events.action import (
+    AgentSummarizeAction,
+)
+from opendevin.llm.messages import Message
+
+from .prompts import (
+    MESSAGE_SUMMARY_WARNING_FRAC,
+    SUMMARY_PROMPT_SYSTEM,
+    parse_summary_response,
+)
+
+
+class CondenserMixin:
+    """Condenses a group of condensable messages as done by MemGPT."""
+
+    def condense(
+        self,
+        messages: list[Message],
+    ):
+        # Start past the system message, and example messages.,
+        # and collect messages for summarization until we reach the desired truncation token fraction (eg 50%)
+        # Do not allow truncation  for in-context examples of function calling
+        token_counts = [
+            self.get_token_count([message])  # type: ignore
+            for message in messages
+            if message.condensable
+        ]
+        message_buffer_token_count = sum(token_counts)  # no system and example message
+
+        desired_token_count_to_summarize = int(
+            message_buffer_token_count * self.config.message_summary_trunc_tokens_frac  # type: ignore
```

Comment:
This mechanism of choosing how much to summarize is nice! The result is that from 11.8k tokens it tries the first condense with 5.8k, and I think that will work well. If it was too close to the upper limit it might risk a fail, but this is almost half at these numbers, and that's safe enough for common use cases.